### ISSUE TITLE
Spherical harmonics parametrization

### ISSUE THREAD

**Issue by VladimirYugay** [wanmeihuali/taichi_3d_gaussian_splatting#150]
In the original splatting [repo](https://github.com/graphdeco-inria/gaussian-splatting), spherical harmonics parameters are represented as `features_rest` of shape (n, 15, 3) and `features_dc` of shape (n, 1, 3) where `n` is the number of points. 

Here the feature vector is of shape 56 by default. First 4 dimensions for rotation, then 3 dimensions for the scale, 1 dimensions for alpha.

My question is, how can I convert the optimized features from the original repo to match the features that you have preserving the correct order?

**Comment by jb-ye** [comment]
I'm afraid matching the order of features isn't enough. I found that the official repo was computing color by `max(0, <SH, dir> + 0.5)`, while this unofficial code compute color using `sigmoid(<SH, dir>)`. 

I also have this question for the @wanmeihuali : what is the design decision to use sigmoid?

**Comment by wanmeihuali** [comment]
Hi @jb-ye, you are right. there are some differences between the official repo and this repo. Consider the volume rendering formula:

$$
C = \sum_{i=0}^{N-1} \alpha_i c_i \prod_{j=0}^{i-1} (1 - \alpha_j)
$$

$c_i$ is the color of each Gaussian, and in most NeRF implementations, its range is $[0, 1]$. When I implemented this repo, because the official repo had not been released, I followed the common practice and used sigmoid to map the range of $c_i$ to $[0, 1]$. However, the official repo uses $max(0, c_i + 0.5)$ to map the range of $c_i$ to $[0, \infty]$, which cause possible overflow in the volume rendering formula, and they clip the pixel color to $[0, 1]$ after the volume rendering formula. I've tried to switch to the official repo's color mapping method, but the result gets worse, so I keep the sigmoid method. Switching between these two methods shall be easy(less than 20 lines of code).

@VladimirYugay Please see https://github.com/wanmeihuali/taichi_3d_gaussian_splatting/blob/2cea5deb3940079efe917ad32a07c8390907507a/benchmark/inference_benchmark.py#L65 for how to import model trained by official implementation, Also you need to modify the color formula a bit.

**Comment by oliver-batchelor** [comment]
I also have been trying to load point clouds from the original repo and try them with the Taichi version. 

I found two things - the adjusted color calculation (below), and the quaternion order was different. 

One thing I haven't been able to figure out is some weird blocking artefacts, they look like specific Gaussians are causing some kind of overflow/NaN or something in the surrounding tiles. I checked and there's no inf or nan in the input data, so I'm not sure what's causing that.


![image](https://github.com/wanmeihuali/taichi_3d_gaussian_splatting/assets/1128677/e0e203ff-2b83-4599-9429-324c2b5eaf9a)


    ```
@ti.func
    def get_color_by_ray(
        self,
        ray_origin: ti.math.vec3,
        ray_direction: ti.math.vec3,
    ) -> ti.math.vec3:
        o = ray_origin
        d = ray_direction
        # TODO: try other methods to get the query point for SH, e.g. the intersection point of the ray and the ellipsoid
        
        rgb = ti.math.vec3(*[SphericalHarmonics(c).evaluate(d) 
                              for c in [self.color_r, self.color_g, self.color_b]])
        return ti.math.clamp(rgb + 0.5, 0, 1)
            # return ti_sigmoid(rgb)
```

**Comment by jb-ye** [comment]
@oliver-batchelor Latest commit, you can find sample code to order SH to match the one used by official GS. https://github.com/wanmeihuali/taichi_3d_gaussian_splatting/blob/main/taichi_3d_gaussian_splatting/GaussianPointCloudScene.py#L148

**Comment by VladimirYugay** [comment]
@wanmeihuali Thanks, that script is helpful! 

Can you elaborate on `Also you need to modify the color formula a bit`?

I was able to load all the parameters and render the image. However, my image looks pale (colors are not so bright), and I think that's because I need to modify the color formula somehow.

**Comment by oliver-batchelor** [comment]
> @wanmeihuali Thanks, that script is helpful!
> 
> Can you elaborate on `Also you need to modify the color formula a bit`?
> 
> I was able to load all the parameters and render the image. However, my image looks pale (colors are not so bright), and I think that's because I need to modify the color formula somehow.

See the code above - the difference is  

taichi version:
`sigmoid(sh(dir, params)) `

3dgs original:
`max(0,  0.5 + sh(dir, params))`

**Comment by VladimirYugay** [comment]
@oliver-batchelor  I see, but it's not that straightforward for me. 

Here's what I tried:

1. Take the features from 3dgs original, and do `features[features != 0] -= 0.5` which should give the the original `sh` which can be decoded to the color values using sigmoid. It didn't work and my image is still blurry

2. I tried precomputing color in 3dgs original using Sigmoid, but it crashes with some memory errors. 

3.  Changing color to `max(0, 0.5 + sh)` from `sigmoid` in taichi also seems a little bit cumbersome - I'd need to adjust derivatives, etc.

Is there any easier way to do it or maybe I'm missing something?

**Comment by VladimirYugay** [comment]
@oliver-batchelor, but I'd also need to change the backward pass if I want to optimize the taichi representation with changed activation functions. Or am I missing something?

**Comment by oliver-batchelor** [comment]
Yeah, you would. I have only tried displaying them at the moment.

**Comment by VladimirYugay** [comment]
@wanmeihuali  I'm not familiar with taichi, and I was wondering whether adding this will also require the change of the gradient for the backward pass:

```
  r = SphericalHarmonics(self.color_r).evaluate(d)
  r = - ti.math.log(1 / (ti.math.clamp(r + 0.5, 0, 1)) - 1)
  r_normalized = ti_sigmoid(r)
```
I found that with this transformation the features from the original report become compatible with the ones in taichi

**Comment by wanmeihuali** [comment]
> @wanmeihuali I'm not familiar with taichi, and I was wondering whether adding this will also require the change of the gradient for the backward pass:
> 
> ```
>   r = SphericalHarmonics(self.color_r).evaluate(d)
>   r = - ti.math.log(1 / (ti.math.clamp(r + 0.5, 0, 1)) - 1)
>   r_normalized = ti_sigmoid(r)
> ```
> 
> I found that with this transformation the features from the original report become compatible with the ones in taichi

@VladimirYugay  For backward, it should be here:
https://github.com/wanmeihuali/taichi_3d_gaussian_splatting/blob/24471486c474c20f9c59607dca078514ab279fc7/taichi_3d_gaussian_splatting/GaussianPoint3D.py#L361-L370

if we remove sigmoid and uses a relu(x+0.5) in stead, then the jocabian shall be the original value when x>0 and 0 then x <=0
